src/julia_to_gap.jl

```diff
@@ -93,37 +99,39 @@ function julia_to_gap(x::Rational{T}) where {T<:Integer}
             return -Globals.infinity
         end
     end
-    numer = julia_to_gap(numer_julia)
-    denom = julia_to_gap(denom_julia)
+    numer = julia_to_gap_internal(numer_julia, cache, recursive)
+    denom = julia_to_gap_internal(denom_julia, cache, recursive)
     return Wrappers.QUO(numer, denom)
 end
 
 ## Floats
-julia_to_gap(x::Float64) = NEW_MACFLOAT(x)
-julia_to_gap(x::Float32) = NEW_MACFLOAT(Float64(x))
-julia_to_gap(x::Float16) = NEW_MACFLOAT(Float64(x))
+GAP.@install GapObj(x::Float64) = NEW_MACFLOAT(x)
+GAP.@install GapObj(x::Float32) = NEW_MACFLOAT(Float64(x))
+GAP.@install GapObj(x::Float16) = NEW_MACFLOAT(Float64(x))
 
 ## Chars
-julia_to_gap(x::Char) = CharWithValue(Cuchar(x))
+GAP.@install GapObj(x::Char) = CharWithValue(Cuchar(x))
 
 ## Strings and symbols
-julia_to_gap(x::AbstractString) = MakeString(string(x))
-julia_to_gap(x::Symbol) = MakeString(string(x))
-
-## Generic caller for optional arguments
-julia_to_gap(obj::Any; recursive::Bool) = julia_to_gap(obj)
-julia_to_gap(obj::Any, recursion_dict::IdDict{Any,Any}; recursive::Bool = true) = julia_to_gap(obj)
+GAP.@install GapObj(x::AbstractString) = MakeString(string(x))
+GAP.@install GapObj(x::Symbol) = MakeString(string(x))
 
 ## Arrays (including BitVector)
-function julia_to_gap(
+function julia_to_gap_internal(
     obj::AbstractVector{T},
-    recursion_dict::IdDict{Any,Any} = IdDict();
-    recursive::Bool = false,
+    recursion_dict::GapCacheDict,
+    recursive::Bool,
 ) where {T}
 
+    if recursion_dict !== nothing && haskey(recursion_dict, obj)
+      return recursion_dict[obj]
+    end
     len = length(obj)
     ret_val = NewPlist(len)
     if recursive
+        if recursion_dict === nothing
+          recursion_dict = RecDict()
```

Comment:
Another optimization note for future iterations on this: if we are given e.g. a `Vector{Int16}` and asked to recursively convert it to a `GapObj`, then we really don't need the `recursion_dict`, as there is no point storing the `Int16` values in it.

To get rid of such things, my "dual" code in PR #777 does this:
```
    if rec_dict === nothing && _needs_conversion_tracking(T)
        rec_dict = RecDict()
    end
```
where `_needs_conversion_tracking(T)` returns `true` if `T` is a type where the conversion results should be tracked.

Such an approach interacts nicely with my suggestion to "split" `GAP.@install`:
- `GAP.@install`  (or whatever it is called) could also produce a `_needs_conversion_tracking` method returning `false`
- `GAP.@install_rec`  (or whatever it is called) could also produce a `_needs_conversion_tracking` method returning `true`


Note: I guess we should have two different `_needs_conversion_tracking` functions, with different names, for the two direction: to and from Julia. Perhaps they'll end up identical but this is not a priori clear to me? Anyway there would be no harm in having two for the moment.